### ISSUE TITLE
Rename `Composite` and `Variant` type defs

### ISSUE THREAD

**Issue by ascjones** [paritytech/scale-info#93]
@Robbepop and I went back and forth over this originally and settled on Composite and Variant as non-rust specific names for `struct` and `enum` types.

However I think this naming is slightly awkward in places, especially when writing code for a "variant" type which itself contains "variants". 

So I think we should just use the `enum` and `struct` terminology, people understand what they are and the code will be easier to read/write.

**Comment by dvdplm** [comment]
If we ever were to support unions, `struct` would be an awkward name (but I don't think we are going to support unions any time soon).

**Comment by jacogr** [comment]
100% on-board with this :) As an implementer it is actually quite awkward when starting to ue these since the terms are completely non-standard to what we use elsewhere. Would also get it in "now-ish" since, well, we have quite a large dependency on this now in metadata and it is the right time to break things.

**Comment by ascjones** [comment]
@Robbepop is strongly in favour of the current names, so we would need to persuade him.

**Comment by jacogr** [comment]
It is fine to keep if that s the way the API is meant to be. 

The undeniable point is that it is absolutely terrible from an external API usability perspective and actually knowing what these are meant to represent.

After seeing this, I've actually started a PR to rename them in the JS API (with parsing aliases for e.g. JSON inputs), since these are now going to start being front-and-center and as exposed just doesn't reflect what it is meant to. Rather than explaining over-and-over in issues (aka what does this mean in the metadata) renaming the mappings is just easier. 

So at least on the JS API types side these will be more usable and maintainable without having a scale PhD, which solves the issue.

**Comment by Robbepop** [comment]
> So at least on the JS API types side these will be more usable and maintainable without having a scale PhD, which solves the issue.

`Composite` and `Variant` are already compromises coming from `Product` and `Sum` type names which are more of the "PhD" terms for the same.
What are the names your JS API proposes for those now?

**Comment by jacogr** [comment]
Since everybody looks at the Rust types anyway to grok in/out in Substrate, attempting to match the definition closer of the underlying types they refer to.

- Struct - as suggested above (it is not 100%, since unnamed can it refer to Tuple-like sequences of fields)
- Enum - as suggested above (mostly ok, although it also matches in actual use, enums that are used in sets, e.g. IdentityFields)
- Vec - additionally replacing sequence
- Array - this is to be replaced, no idea as of "what" (it has specific meaning more akin to Vec as currently named, however it is a fixed length sequence of items)

YMMV. (The above will gets some tweaks over the next day or two, real focus is making sure it works end-to-end and finding any remaining nigglies, renaming is easy pre-merge into Substrate)

**Comment by Robbepop** [comment]
> Since everybody looks at the Rust types anyway to grok in/out in Substrate, attempting to match the definition closer of the underlying types they refer to.
> 
>     * Struct - as suggested above (it is not 100%, since unnamed can it refer to Tuple-like sequences of fields)
> 
>     * Enum - as suggested above (mostly ok, although it also matches in actual use, enums that are used in sets, e.g. IdentityFields)
> 
>     * Vec - additionally replacing sequence
> 
>     * Array - this is to be replaced, no idea as of "what" (it has specific meaning more akin to Vec as currently named, however it is a fixed length sequence of items, here even the base SCALE implementation on the JS API is horrible named, e.g `VecFixed` which is less than optimal, but had no better ideas at the time and it sort-of stuck around)
> 
> 
> YMMV. (The above will gets some tweaks over the next day or two, real focus is making sure it works end-to-end and finding any remaining nigglies, renaming is easy pre-merge into Substrate)

Let me quickly go over why I think those names are not a good idea.

- `Struct` as you said yourself this kind of confuses people by making them think about Rust structs without going beyond that tuple structs and tuples are included as well. You made this point yourself though.
- `Enum`: Many people coming from C/Java/C#/C++ won't recon `enum` as something that may hold data. Therefore for all of those people this is again a bad naming and confusing. Even some Rust people think that `enum` was a bad name for the construct.
- `Vec` instead of `Sequence`: This would give people the wrong impression that all the different collections that are serialized similar to a vector as in fact just vectors. So sequence is again a more neutral term that covers more of what is actually happening.
- `Array`: I agree `Array` is not a good term either but `Array` in C/C++/Rust usually have fixed lengths so I thought it was the least bad option with a name that is bearable to type in.

**Comment by jacogr** [comment]
We are not in C++ - we are coding to SCALE with the de-facto implementation in Rust.

The naming as you have fought for is utterly confusing. Completely and utterly. (And I'm putting it lightly)

**Comment by dvdplm** [comment]
I for one don't mind the current naming scheme. 

Yes, it took some getting used to, but the effort that Andrew and Robin put in to making the whole library as generic as possible has paid off greatly in other circumstances and I don't think we should rush into Rusty names just because that's what we and many in our current audience is used to.

I agree with @Robbepop that "enum" is a slightly unfortunate name even for Rust itself; switching out "composite" (which I don't love) to "struct" would be terrible imo. "sequence" is as good of a name as it gets.

Substrate requires practicians to learn a metric fuckton of terminology already; I'm not convinced then learning what "composite" and "variant" maps to is going to be a big problem.

**Comment by jacogr** [comment]
I love what Andrew did here and I am certainly a fan of his work. But we dropped the ball since we try to not map to what we have as input. 

All I can say is - actually use the libs and then come back.  Andrew has, he logged this issue. I have, I am supporting the direction he wants to take it in.  Sadly we are blocked by an idealized view way people who don’t actually code against the stuff. 

To each their own. Next time, hopefully we do it right.
